LinkRelAttribute doesn't tokenize correctly

annevk opened the pull request:
#### 13d238d72b7e3987830641bdc7f85ade40159ce1
<pre>
LinkRelAttribute doesn&apos;t tokenize correctly
<a href="https://bugs.webkit.org/show_bug.cgi?id=285508">https://bugs.webkit.org/show_bug.cgi?id=285508</a>

Reviewed by Ryosuke Niwa.

Make LinkRelAttribute use SortedArrayMap as a backing store for the
link types so we can reuse the logic and thereby make the overall setup
less error prone and fix several bugs in the process.

Also correct tokenizing by recognizing all ASCII whitespace instead of
just newlines and spaces.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/document-metadata/the-link-element/link-rel-attribute-tokenization-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/document-metadata/the-link-element/link-rel-attribute-tokenization.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/document-metadata/the-link-element/resources/w3c-import.log:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/document-metadata/the-link-element/w3c-import.log:
* Source/WebCore/html/LinkRelAttribute.cpp:
(WebCore::linkTypes):
(WebCore::findLinkType):
(WebCore::LinkRelAttribute::LinkRelAttribute):
(WebCore::LinkRelAttribute::isSupported):

Canonical link: <a href="https://commits.webkit.org/288639@main">https://commits.webkit.org/288639@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/da46d5454e2fe933a4944e713c07b07d4032a5ce

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83904 "12 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3522 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38205 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/88978 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/34912 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/85989 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3613 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11489 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65261 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23094 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86950 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2682 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76235 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45553 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2612 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30457 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/33961 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73601 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31211 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90354 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11169 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8079 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73706 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11393 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72064 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72924 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17210 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/15898 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/2503 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12996 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11121 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16593 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/10969 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14445 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/12741 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/html/LinkRelAttribute.cpp:
```diff
@@ -40,67 +40,85 @@
 namespace WebCore {
 
+using LinkRelAttributeHandler = void (*)(LinkRelAttribute& relAttribute);
+using LinkRelAttributeMap = HashMap<ASCIILiteral, LinkRelAttributeHandler>;
```
Please use `MemoryCompactLookupOnlyRobinHoodHashMap` instead.

rniwa commented on Source/WebCore/html/LinkRelAttribute.cpp:
```diff
@@ -40,67 +40,85 @@
+using LinkRelAttributeMap = HashMap<ASCIILiteral, LinkRelAttributeHandler>;
+static LinkRelAttributeMap createLinkRelAttributeMap()
+{
+    static NeverDestroyed<LinkRelAttributeMap> linkRelAttributeMap = LinkRelAttributeMap({
```
Since there are so few items, binary search is probably faster than hash table lookup.

rniwa replied:
See isPotentialCustomElementNameCharacter in Document.cpp for an example of binary search.

annevk replied:
There is a single place where this can be implemented now (findLinkType), but it's not done as I wasn't quite sure about the best approach. I did look at that method and also did some research into writing custom a compare method as would be needed here, but it seems quite involved and I couldn't easily find precedent in the codebase.

I'm kinda hoping that since the performance should be about equivalent with what we have today we can take this patch as-is and separately figure out if we want to add a binary search later on.

rniwa commented on Source/WebCore/html/LinkRelAttribute.cpp:
```diff
@@ -40,67 +40,85 @@
+            if (found)
+                continue;
+
+            if (document.settings().linkPreconnectEnabled() && equalLettersIgnoringASCIICase(possibleLinkType, "preconnect"_s))
```
Since there are no chance of these entries ever matching other entires in the map,
we might as well as use the same lookup table for all these runtime enabled features,
and check the runtime enabled-ness optionally:

```cpp
auto found = ([&] {
    for (auto& entry : linkTypeTable) {
        if (equalLettersIgnoringASCIICase(entry.name, possibleLinkType) && entry.isEnabled()) {
            entry.updateRelAttribute(*this);
            return true;
        }
    }
    return false;
})();
```

cdumez commented on Source/WebCore/html/LinkRelAttribute.cpp:
```diff
@@ -40,82 +41,79 @@
+
+static std::optional<const LinkType> findLinkType(StringView rel)
+{
+    for (auto linkType : linkTypes) {
```
I think this is the kind of thing we usually use `SortedArrayMap` (+ `ComparableLettersLiteral`) for. cc @darinadler

annevk replied:
Interesting! That does have this comment at the top indicating it might not be a win given the current size:
```
class SortedArrayBase {
protected:
    // Some informal empirical tests indicate that arrays shorter than this are faster to
    // search with linear search than with binary search. Even if we don't get this threshold
    // exactly right, it's helpful for both performance and code size to use linear search at
    // least for very small arrays, and important for performance to make sure that we use
    // binary search for much larger ones.
    static constexpr size_t binarySearchThreshold = 20;
};
```
But if I'm correct about that I should probably at least add a comment indicating this should be changed to `SortedArrayMap` once we hit that number.

cdumez replied:
I don't think it matters, it is still nicer than doing your own linear search with case insensitive matching I think.

annevk replied:
That is fair. Certainly good to rely on more fundamental primitives. (To be clear, this is now resolved, but leaving this unresolved so Darin can continue to see the context.)

darinadler replied:
One of the best features of `SortedArrayMap` is that it automatically starts using the better algorithm as the table gets longer and does a linear search when that's more efficient. So I suggest using it without thinking twice about the size of the table

rniwa replied:
@darinadler : Maybe the comment in SortedArrayBase should be updated accordingly?

darinadler replied:
Sure, I’ll do that.

annevk committed:
LinkRelAttribute doesn't tokenize correctly
https://bugs.webkit.org/show_bug.cgi?id=285508

Reviewed by Ryosuke Niwa.

Make LinkRelAttribute use SortedArrayMap as a backing store for the
link types so we can reuse the logic and thereby make the overall setup
less error prone and fix several bugs in the process.

Also correct tokenizing by recognizing all ASCII whitespace instead of
just newlines and spaces.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/document-metadata/the-link-element/link-rel-attribute-tokenization-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/document-metadata/the-link-element/link-rel-attribute-tokenization.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/document-metadata/the-link-element/resources/w3c-import.log:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/document-metadata/the-link-element/w3c-import.log:
* Source/WebCore/html/LinkRelAttribute.cpp:
(WebCore::linkTypes):
(WebCore::findLinkType):
(WebCore::LinkRelAttribute::LinkRelAttribute):
(WebCore::LinkRelAttribute::isSupported):

Canonical link: https://commits.webkit.org/288639@main
